src/swap/defi/rango.ts

```diff
@@ -0,0 +1,542 @@
+    } else {
+      toContractAddress = toToken?.networkLocation?.contractAddress
+    }
+
+    if (fromContractAddress == null || toContractAddress == null) {
+      throw new SwapCurrencyError(swapInfo, request)
+    }
+
+    // Do not support transfer between same assets
+    if (
+      fromWallet.currencyInfo.pluginId === toWallet.currencyInfo.pluginId &&
+      request.fromCurrencyCode === request.toCurrencyCode
+    ) {
+      throw new SwapCurrencyError(swapInfo, request)
+    }
+
+    const fromAddress = await getAddress(fromWallet)
+    const toAddress = await getAddress(toWallet)
+
+    const fromMainnetCode =
+      MAINNET_CODE_TRANSCRIPTION[fromWallet.currencyInfo.pluginId]
+    const toMainnetCode =
+      MAINNET_CODE_TRANSCRIPTION[toWallet.currencyInfo.pluginId]
+
+    if (fromMainnetCode === '' || toMainnetCode === '') {
+      throw new SwapCurrencyError(swapInfo, request)
+    }
+
+    const now = Date.now()
+    if (
+      now - exchangeInfoLastUpdate > EXCHANGE_INFO_UPDATE_FREQ_MS ||
+      exchangeInfo == null
+    ) {
+      try {
+        const exchangeInfoResponse = await promiseWithTimeout(
+          fetchInfo(fetchCors, `v1/exchangeInfo/${appId}`)
+        )
+
+        if (exchangeInfoResponse.ok === true) {
+          exchangeInfo = asExchangeInfo(await exchangeInfoResponse.json())
+          exchangeInfoLastUpdate = now
+        } else {
+          // Error is ok. We just use defaults
+          const text: string = await exchangeInfoResponse.text()
+          log.warn(
+            `Error getting info server exchangeInfo. Using defaults... Error: ${text}`
+          )
+        }
+      } catch (e: any) {
+        log.warn(
+          'Error getting info server exchangeInfo. Using defaults...',
+          e.message
+        )
+      }
+    }
+
+    const fromSymbol =
+      rangoTokens?.[fromMainnetCode]?.[fromContractAddress.toLowerCase()]
+    const toSymbol =
+      rangoTokens?.[toMainnetCode]?.[toContractAddress.toLowerCase()]
+
+    if (fromSymbol === undefined || toSymbol === undefined) {
+      throw new Error('Rango could not proceed with the exchange')
+    }
+
+    if (exchangeInfo != null) {
+      const { rango } = exchangeInfo.swap.plugins
+      rangoServers = rango?.rangoServers ?? rangoServers
+    }
+
+    const swapParameters = {
+      apiKey: rangoApiKey,
+      from: assetToString({
+        blockchain: fromMainnetCode,
+        address: fromContractAddress,
+        symbol: fromSymbol
+      }),
+      to: assetToString({
+        blockchain: toMainnetCode,
+        address: toContractAddress,
+        symbol: toSymbol
+      }),
+      fromAddress: fromAddress,
+      toAddress: toAddress,
+      amount: nativeAmount,
+      disableEstimate: true,
+      slippage: DEFAULT_SLIPPAGE,
+      ...(referrerAddress !== undefined &&
+        referrerAddress !== '' &&
+        referrerFee !== undefined &&
+        referrerFee !== '' && { referrerAddress, referrerFee })
```

Comment:
bizarre syntax we don't use. Declare referrerAddress and referrerFee above and put into `swapParameters`